File: core/src/com/github/servb/antonbird/helper/AssetLoader.java
/*
 * Copyright 2018 SerVB.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.servb.antonbird.helper;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.audio.Sound;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.Texture.TextureFilter;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.TextureRegion;

/**
 *
 * @author SerVB
 */
public final class AssetLoader {

    public static final int DIE_SOUND_COUNT = 4;

    public static Texture texture;
    public static TextureRegion bg, grass;

    public static Animation birdAnimation;
    public static TextureRegion bird, birdDown, birdUp;

    public static TextureRegion skullUp, skullDown, bar;

    public static Sound[] dead;
    public static Sound flap;
    public static Sound coin;

    public static BitmapFont font, shadow;

    public static final void load() {
        texture = new Texture(Gdx.files.internal("data/texture.png"));
        texture.setFilter(TextureFilter.Nearest, TextureFilter.Nearest);

        bg = new TextureRegion(texture, 0, 0, 136, 43);
        bg.flip(false, true);

        grass = new TextureRegion(texture, 0, 43, 143, 11);
        grass.flip(false, true);

        birdDown = new TextureRegion(texture, 136, 0, 17, 12);
        birdDown.flip(false, true);

        bird = new TextureRegion(texture, 153, 0, 17, 12);
        bird.flip(false, true);

        birdUp = new TextureRegion(texture, 170, 0, 17, 12);
        birdUp.flip(false, true);

        birdAnimation = new Animation(0.06f, birdDown, bird, birdUp);
        birdAnimation.setPlayMode(Animation.PlayMode.LOOP_PINGPONG);

        skullUp = new TextureRegion(texture, 192 + 24, 0, 24, 14);
        skullUp.flip(false, true);
        skullDown = new TextureRegion(texture, 192, 0, 24, 14);
        skullDown.flip(false, true);

        bar = new TextureRegion(texture, 136, 16, 22, 3);
        bar.flip(false, true);

        dead = new Sound[DIE_SOUND_COUNT];
        for (int i = 0; i < dead.length; ++i) {
            dead[i] = Gdx.audio.newSound(Gdx.files.internal("data/die" + i + ".wav"));
        }

        flap = Gdx.audio.newSound(Gdx.files.internal("data/flap.wav"));
        coin = Gdx.audio.newSound(Gdx.files.internal("data/coin.wav"));

        font = new BitmapFont(Gdx.files.internal("data/text.fnt"));
        font.getData().setScale(.25f, -.25f);
        shadow = new BitmapFont(Gdx.files.internal("data/shadow.fnt"));
        shadow.getData().setScale(.25f, -.25f);
    }

    public static final void dispose() {
        texture.dispose();

        for (int i = 0; i < dead.length; ++i) {
            dead[i].dispose();
        }
        flap.dispose();
        coin.dispose();

        font.dispose();
        shadow.dispose();
    }
}

File: core/src/com/github/servb/antonbird/helper/AssetLoader.kt
package com.github.servb.antonbird.helper

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.Texture.TextureFilter
import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.TextureRegion

object AssetLoader {

    private const val DIE_SOUND_COUNT = 4

    private lateinit var texture: Texture
    lateinit var bg: TextureRegion
    lateinit var grass: TextureRegion

    lateinit var birdAnimation: Animation<TextureRegion>
    lateinit var bird: TextureRegion
    lateinit var birdDown: TextureRegion
    lateinit var birdUp: TextureRegion

    lateinit var skullUp: TextureRegion
    lateinit var skullDown: TextureRegion
    lateinit var bar: TextureRegion

    lateinit var dead: Array<Sound>
    lateinit var flap: Sound
    lateinit var coin: Sound

    lateinit var font: BitmapFont
    lateinit var shadow: BitmapFont

    fun load() {
        texture = Texture(Gdx.files.internal("data/texture.png")).apply {
            setFilter(TextureFilter.Nearest, TextureFilter.Nearest)
        }

        bg = TextureRegion(texture, 0, 0, 136, 43).apply {
            flip(false, true)
        }

        grass = TextureRegion(texture, 0, 43, 143, 11).apply {
            flip(false, true)
        }

        birdDown = TextureRegion(texture, 136, 0, 17, 12).apply {
            flip(false, true)
        }

        bird = TextureRegion(texture, 153, 0, 17, 12).apply {
            flip(false, true)
        }

        birdUp = TextureRegion(texture, 170, 0, 17, 12).apply {
            flip(false, true)
        }

        birdAnimation = Animation(0.06f, birdDown, bird, birdUp).apply {
            playMode = Animation.PlayMode.LOOP_PINGPONG
        }

        skullUp = TextureRegion(texture, 192 + 24, 0, 24, 14).apply {
            flip(false, true)
        }
        skullDown = TextureRegion(texture, 192, 0, 24, 14).apply {
            flip(false, true)
        }

        bar = TextureRegion(texture, 136, 16, 22, 3).apply {
            flip(false, true)
        }

        dead = Array(DIE_SOUND_COUNT) { i -> Gdx.audio.newSound(Gdx.files.internal("data/die$i.wav")) }
        flap = Gdx.audio.newSound(Gdx.files.internal("data/flap.wav"))
        coin = Gdx.audio.newSound(Gdx.files.internal("data/coin.wav"))

        font = BitmapFont(Gdx.files.internal("data/text.fnt")).apply {
            data.setScale(0.25f, -0.25f)
        }
        shadow = BitmapFont(Gdx.files.internal("data/shadow.fnt")).apply {
            data.setScale(0.25f, -0.25f)
        }
    }

    fun dispose() {
        texture.dispose()


        dead.forEach { it.dispose() }
        flap.dispose()
        coin.dispose()

        font.dispose()
        shadow.dispose()
    }
}
